feat(act 1): activity 1 - clean code basics (#11)
pyOpenSci/lessons: clean-modular-code/activity-1/clean-code-activity-1.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

(clean-code-activity-1)=
# Clean, Modular Code: Activity 1

> Writing clean, modular code takes practice but is a habit worth building. Over time as you incorporate clean code strategies that improve code quality and maintainability, these strategies will become second nature. This, in turn, will make your code easier to read, maintain, and share with others. It will also make it easier for others (and your future self) to use. 

Reviewer: ```suggestion
Writing clean, modular code takes practice but is a habit worth building. Over time as you incorporate clean code strategies that improve code quality and maintainability, these strategies will become second nature. This, in turn, will make your code easier to read, maintain, and share with others. It will also make your work easier for others (and your future self) to use. 
```

In this exercise, you'll focus on three key clean coding strategies:

1. **Use expressive names**: Assign meaningful names to all variables and functions to make your code more readable. [Learn more about expressive names.](../python-expressive-code.md)
2. **Follow PEP8 guidelines**: Adhere to [PEP8 Python code style rules](../python-pep-8.md), including proper spacing and naming conventions, to maintain a consistent and readable codebase.
3. **Identify opportunities to make your code DRY (Don't Repeat Yourself)**: In this activity, you will use pseudocode to pinpoint areas where the code can be simplified and made DRY. In the next activity, you will implement DRY best practices using loops and functions. 

By practicing these strategies, you are well on your way to writing clean, efficient, and maintainable code.

## Activity data 

> Below is some code that processes cross-ref citation data for JOSS publications. The data were pulled directly from the cross-ref API but then modified with some specific features to help you learn when completing the activities.

Reviewer: ```suggestion
Today's activity begins with some code that processes cross-ref citation data for JOSS publications. The data were pulled directly from the cross-ref API but then modified with specific features to help you learn when completing the activities.
```

## Your goal 
Your goal is to take the code below and turn it into a script that has the following characteristics:

* The code uses clean, expressive naming conventions
* the code follows the PEP 8 style guide

While you will implement the items below in the next activity, you will create pseudocode in this activity to identify:

* How the code could be more DRY. Are elements repeated? 

> * How the code could be more modular. 

Reviewer: ```suggestion
* How the code could be made more modular. 
```

### If you want to use an LLM to support your learning 

> The cleanup steps discussed above are things an LLM can help you with. However, remember that if you use LLMs, they likely will not always give you the right answer. In fact more often than not, part of the answer is wrong. This means that you need to have a keen eye to catch issues in your code.

Reviewer: ```suggestion
The cleanup steps discussed above are things an LLM can help you with. However, remember that if you use LLMs, they likely will not always give you the right answer. In fact more often than not, part of the answer is wrong. This means that you need to have a keen eye to catch issues in LLM-generated code.
```

If you are using a LLM: 

* provide it with descriptive, leading prompts that allow it to perform the task better. So you might write:

> > Make the variables in the code below.

Author: fixed! thanks!
